sql/src/main/java/io/crate/operation/scalar/ArrayMergeFunction.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Licensed to CRATE Technology GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.operation.scalar;
+
+import io.crate.metadata.*;
+import io.crate.operation.Input;
+import io.crate.planner.symbol.Function;
+import io.crate.planner.symbol.Literal;
+import io.crate.planner.symbol.Symbol;
+import io.crate.types.DataType;
+import io.crate.types.ArrayType;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+
+public abstract class ArrayMergeFunction extends Scalar<Object[], Object> {
+
+    public static final String NAME = "array_merge";
+    private FunctionInfo functionInfo;
+    private static ArrayType arrayType;
+
+    private static FunctionInfo createInfo(List<DataType> types) {
+        arrayType = (ArrayType) types.get(0);
+        return new FunctionInfo(new FunctionIdent(NAME, types), arrayType);
+    }
+
+    public static void register(ScalarFunctionModule module) {
+        module.register(NAME, new Resolver());
+    }
+
+    protected ArrayMergeFunction(FunctionInfo functionInfo) {
+        this.functionInfo = functionInfo;
+    }
+
+    @Override
+    public FunctionInfo info() {
+        return functionInfo;
+    }
+
+    @Override
+    public Symbol normalizeSymbol(Function symbol) {
+        if (anyNonLiterals(symbol.arguments())) {
+            return symbol;
+        }
+
+        Input[] argList = new Input[symbol.arguments().size()];
+        int counter = 0;
+        for(Symbol argSymbol : symbol.arguments()){
+            Input input  = (Input) argSymbol;
+            argList[counter++] = input;
+        }
+
+        return Literal.newLiteral(evaluate(argList), arrayType);
+    }
+
+    private static class MergeFunction extends ArrayMergeFunction {
+
+        protected MergeFunction(FunctionInfo functionInfo) {
+            super(functionInfo);
+        }
+
+        @Override
+        public Object[] evaluate(Input[] args) {
+            List<Object> resultList = new ArrayList<>();
```

Comment:
instead of using an ArrayList which copies and increases its internal array you could get the sum of the sizes of the argument arrays in a first pass, and in a second copy the arg objects over into that newly created array.